### PR TITLE
Handle edits that scroll up due to hiding the horizontal scrollbar

### LINKED ISSUE #15263: Uncaught Error: Requested measurement of a line component that is not currently rendered
[Enter steps to reproduce:]

1. ...
2. ...

**Atom**: 1.19.0 x64
**Electron**: 1.6.9
**OS**: Microsoft Windows 10 Enterprise
**Thrown From**: Atom Core

### Stack Trace

Uncaught Error: Requested measurement of a line that is not currently rendered

```
At C:\Users\sarr\AppData\Local\atom\app-1.19.0\resources\app\src\text-editor-component.js:2173

Error: Requested measurement of a line that is not currently rendered
    at horizontalPositionsToMeasure.forEach (~/AppData/Local/atom/app-1.19.0/resources/app/src/text-editor-component.js:2171:29)
    at Map.forEach (native)
    at TextEditorComponent.measureHorizontalPositions (~/AppData/Local/atom/app-1.19.0/resources/app/src/text-editor-component.js:2164:45)
    at TextEditorComponent.measureContentDuringUpdateSync (~/AppData/Local/atom/app-1.19.0/resources/app/src/text-editor-component.js:385:16)
    at TextEditorComponent.updateSync (~/AppData/Local/atom/app-1.19.0/resources/app/src/text-editor-component.js:276:18)
    at TextEditorComponent.didMouseWheel (~/AppData/Local/atom/app-1.19.0/resources/app/src/text-editor-component.js:1509:43)
```

### Commands

```
     -1:18.7.0 core:save (input.hidden-input)
     -1:14.6.0 tree-view:duplicate (span.name.icon.icon-file-text)
  2x -1:13.8.0 core:move-right (input.hidden-input)
     -1:11.6.0 core:confirm (input.hidden-input)
     -1:10.6.0 core:close (input.hidden-input)
     -1:10 core:select-all (input.hidden-input)
     -1:09.6.0 core:paste (input.hidden-input)
     -1:09.2.0 core:save (input.hidden-input)
  2x -0:54.4.0 core:backspace (input.hidden-input)
     -0:53.2.0 core:save (input.hidden-input)
     -0:47.7.0 editor:select-to-beginning-of-word (input.hidden-input)
     -0:47 core:move-right (input.hidden-input)
  9x -0:46.3.0 core:move-down (input.hidden-input)
     -0:44.5.0 core:move-up (input.hidden-input)
  2x -0:40.7.0 core:undo (input.hidden-input)
     -0:30.7.0 core:save (input.hidden-input)
```

### Non-Core Packages

```
atom-beautify 0.30.4 
atom-css-comb 3.3.1 
language-postcss 1.3.1 
language-vue 0.23.0 
monokai-dark 1.3.0 
seti-ui 1.9.0 
```

### LINKED ISSUE #16289: Overlapping text when cutting folds
### Prerequisites

* [ ] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

Occasionally when there are a lot of folds, cutting folds results in overlapping text such as the following.
folds-overlapping-text

I've managed to reproduce it twice now on https://github.com/atom/language-php/blob/3e45569d8fb3af6747f6e4208b16d17deb577b97/spec/php-spec.coffee while moving heredoc and nowdoc specs into their own `describe`s.

### Steps to Reproduce

1. Open https://github.com/atom/language-php/blob/3e45569d8fb3af6747f6e4208b16d17deb577b97/spec/php-spec.coffee
2. Fold everything and then unfold the top-level describe
3. Try cutting and pasting some folds until it reproduces?

**Expected behavior:** Text to appear in order

**Actual behavior:** Text occasionally overlaps

**Reproduces how often:** Unknown, I'm guessing ~10-20%.

### Versions

```
Atom    : 1.24.0-dev-f76a7a6ed
Electron: 1.6.15
Chrome  : 56.0.2924.87
Node    : 7.4.0
```

### LINKED ISSUE #16710: Uncaught Error: Requested measurement of a line component that is not currently rendered
[Enter steps to reproduce:]

1. ...
2. ...

**Atom**: 1.23.3 x64
**Electron**: 1.6.15
**OS**: Microsoft Windows 8.1 Pro
**Thrown From**: Atom Core

### Stack Trace

Uncaught Error: Requested measurement of a line component that is not currently rendered

```
At C:\Users\musta_000\AppData\Local\atom\app-1.23.3\resources\app\src\text-editor-component.js:2289

Error: Requested measurement of a line component that is not currently rendered
    at horizontalPositionsToMeasure.forEach (~/AppData/Local/atom/app-1.23.3/resources/app/src/text-editor-component.js:2278:29)
    at Map.forEach (native)
    at TextEditorComponent.measureHorizontalPositions (~/AppData/Local/atom/app-1.23.3/resources/app/src/text-editor-component.js:2271:45)
    at TextEditorComponent.measureContentDuringUpdateSync (~/AppData/Local/atom/app-1.23.3/resources/app/src/text-editor-component.js:400:16)
    at scheduler.readDocument (~/AppData/Local/atom/app-1.23.3/resources/app/src/text-editor-component.js:278:20)
    at ViewRegistry.performDocumentUpdate (~/AppData/Local/atom/app-1.23.3/resources/app/src/view-registry.js:245:13)
```

### Commands

```
     -2:38.2.0 snippets:next-tab-stop (input.hidden-input)
     -2:38.2.0 snippets:expand (input.hidden-input)
     -2:38.2.0 editor:indent (input.hidden-input)
     -2:37.6.0 core:save (input.hidden-input)
     -2:30.4.0 snippets:next-tab-stop (input.hidden-input)
     -2:30.4.0 snippets:expand (input.hidden-input)
     -2:30.4.0 editor:indent (input.hidden-input)
  2x -2:29.9.0 core:save (input.hidden-input)
     -1:54.2.0 core:backspace (input.hidden-input)
     -1:53.5.0 core:undo (input.hidden-input)
     -1:53 core:save (input.hidden-input)
     -1:48.6.0 snippets:next-tab-stop (input.hidden-input)
     -1:48.6.0 snippets:expand (input.hidden-input)
     -1:48.6.0 editor:indent (input.hidden-input)
     -1:47.7.0 core:undo (input.hidden-input)
     -1:03.3.0 core:save (input.hidden-input)
```

### Non-Core Packages

```
autocomplete-javascript 0.0.2 
javascript-snippets 1.2.1 
jQuery 1.2.5 
jquery-snippets 11.0.0 
minimap 4.29.7 
teletype 0.7.0 
```

### LINKED ISSUE #16722: Editor sometimes scrambles editor content by rendering different lines over each other
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: https://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: https://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

Sometimes Atom scrambles the editor content by rendering different lines over each other and even scrolling or minimizing/maximizing the editor does not correctly refresh the content.
The problem happens every couple of hours and is completely in-deterministic.
It can only be resolved by closing the editor and starting it again.

2018-02-13 07_33_37-convert js d__mydev_ljs_app_trunk_periscope_build atom

### Steps to Reproduce

Cannot be reproduced

**Expected behavior:** [What you expect to happen]

**Actual behavior:** [What actually happens]

**Reproduces how often:** [What percentage of the time does it reproduce?]

Typically once every couple of hours.

### Versions

Atom    : 1.23.3
Electron: 1.6.15
Chrome  : 56.0.2924.87
Node    : 7.4.0

apm  1.18.12
npm  3.10.10
node 6.9.5 x64
atom 1.23.3
python 2.7.3
git 2.12.2.windows.2
visual studio 2013

Running on Windows 7 x64

### PR DESCRIPTION
**Pull request by Ben3eeE**
## TODO

- [ ] Style the horizontal scrollbar in tests so it's the same height on all platforms
- [x] Check if this fixes https://github.com/atom/atom/issues/16289 

---

**Before:**
before safe mode

**After:**
after dev mode

Fixes https://github.com/atom/atom/issues/16289
Fixes https://github.com/atom/atom/issues/16722
Fixes https://github.com/atom/atom/issues/16710
Fixes https://github.com/atom/atom/issues/15263

### CODE REVIEW

**Comment by Ben3eeE** [comment]
@as-cii @nathansobo appveyor is getting the same failure that I saw locally.

**Comment by Ben3eeE** [comment]
@as-cii I paired with @nathansobo on this. Can you take a look at our proposed change when you have time? 👀 

Setup:
We have an editor that is 1 line tall and the scrollbar and line height are both 17 so the scrollbar is obscuring the entire editor. We are scrolled down exactly one line.

We then delete all content except the first character of the file which causes an update.
Now when the `updateSyncBeforeMeasuringContent` runs we are still believing that we can scroll vertically because the measurement of the horizontal scrollbar is not done until `measureContentDuringUpdateSync` because it's based on the longest line which no longer is true because this measurement is done _after_ the `updateSyncBeforeMeasuringContent` function.

Our proposed change is to restart the frame in scenarios where the scrollbar visibility changes during the update. Maybe you have some other idea on how we could solve this?

We still need to change the spec to style the scrollbars so it's failing on other platforms and probably rebase and clean up the git history.

**Comment by nathansobo** [comment]
@Ben3eeE Hope you don't mind. I made a small clarifying edit to the body of this PR.

File: spec/text-editor-component-spec.js [hunk 1]
```diff
@@ -896,6 +896,39 @@ describe('TextEditorComponent', () => {
...
+      await setEditorHeightInLines(component, 1)
+      await setEditorWidthInCharacters(component, 7)
+
+      element.focus()
+      component.setScrollTop(component.measurements.lineHeight)
+
+      component.scheduleUpdate()
+      await component.getNextUpdatePromise()
+
+      editor.setSelectedBufferRange(new Range(Point(0,1),Point(12,2)))
```
Comment (nathansobo) [inline spec/text-editor-component-spec.js]: It's fine to express a range as nested arrays like `[[0, 1], [12, 2]]`, especially in tests.

File: src/text-editor-component.js [hunk 2]
```diff
@@ -266,9 +266,13 @@ class TextEditorComponent {
     if (useScheduler === true) {
       const scheduler = etch.getScheduler()
       scheduler.readDocument(() => {
-        this.measureContentDuringUpdateSync()
+        const restartFrame = this.measureContentDuringUpdateSync()
         scheduler.updateDocument(() => {
-          this.updateSyncAfterMeasuringContent()
+          if (restartFrame) {
```
Comment (nathansobo) [inline src/text-editor-component.js]: We may need this on the non-scheduled code path as well.

File: src/text-editor-component.js [hunk 3]
```diff
@@ -2118,6 +2125,7 @@ class TextEditorComponent {
       // rendered start row accurately. 😥
       this.populateVisibleRowRange(renderedStartRow)
       this.props.model.setEditorWidthInChars(this.getScrollContainerClientWidthInBaseCharacters())
+      this.derivedDimensionsCache = {}
```
Comment (nathansobo) [inline src/text-editor-component.js]: Did this fix some other failing seed or something?

Comment (Ben3eeE) [reply src/text-editor-component.js]: This is cherry-picked from your PR which was not yet merged when this was opened https://github.com/atom/atom/pull/15828 just to verify that PR wouldn't fix this failure.

**Comment by Ben3eeE** [comment]
@nathansobo Yeah that failure is something Justin has already fixed. We should definitely rebase this.

**Comment by as-cii** [comment]
What's the status of this? Should we go ahead and merge it? It looks reasonable, although seems like tests are failing.

**Comment by Ben3eeE** [comment]
> What's the status of this? Should we go ahead and merge it? It looks reasonable, although seems like tests are failing.

We need to address the review feedback by @nathansobo and style the scrollbars in the test to be the same size as the lineHeight so it's consistently failing on all platforms. It was only failing on appveyor because both the scrollbar and lineHeight was both 17 or something.

The failing test was on master when this PR was created and has already been fixed and everything should be green if we rebase with master.

I won't have time to work on these things anytime soon and will be away for quite a while starting today 🌴

**Comment by Ben3eeE** [comment]
Addressed the review comments by @nathansobo and edited the body of the PR to reflect the current status of this PR.

**Comment by nathansobo** [comment]
This fixes some serious known issues but @Ben3eeE is having trouble coming up with a regression test that works on all platforms. In the interest of not delaying further, I'm going to merge this now and @Ben3eeE is going to follow up with another attempt at a regression test, possibly based on hard coding a failing seed in addition to the randomly generated ones in our randomized test.

**Comment by nathansobo** [comment]
Thanks @Ben3eeE. I feel honored to work with you and am truly thankful for all that you contribute. ❤️ :zap:

**Pull request merged**